File: testdata/fixtures/typescript/CWE-22.ts
// CWE-22: Path Traversal – unsanitized user input flowing into fs file APIs
// allows attackers to read arbitrary files outside the intended directory.
// The sink is fs.readFile / readFileSync / createReadStream; the source is
// request query/params. The safe versions use hard-coded paths.

import fs from "fs";
import { Request } from "express";

function vuln(req: Request) {
  const p = req.query.get("file") as string;
  // ruleid: typescript-path-traversal
  fs.readFile(p, (err, data) => {});
  const q = req.params.get("path") as string;
  // ruleid: typescript-path-traversal
  fs.readFileSync(q);
  const r = req.body.get("r") as string;
  // ruleid: typescript-path-traversal
  fs.createReadStream(r);
  const s = location.search;
  // ruleid: typescript-path-traversal
  fs.promises.readFile(s);
}

function safe(req: Request) {
  // ok: typescript-path-traversal
  fs.readFile("/tmp/fixed.txt", (e, d) => {});
  // ok: typescript-path-traversal
  fs.readFileSync("/tmp/fixed.txt");
  // ok: typescript-path-traversal
  fs.createReadStream("/tmp/fixed.txt");
  // ok: typescript-path-traversal
  fs.promises.readFile("/tmp/fixed.txt");
}

File: testdata/fixtures/clojure/CWE-22.clj
;; CWE-22: Path Traversal – user-controlled path passed to slurp/reader
;; can read arbitrary files. The sink is slurp; safe code uses literals.

(ns fixture.cwe22)

(defn vuln [request]
  (let [p (get-in request [:params :file])]
    ;; ruleid: clojure-path-traversal
    (slurp p))
  (let [q (get-in request [:params :path])]
    ;; ruleid: clojure-path-traversal
    (slurp q))
  (let [r (get-in request [:query-params :id])]
    ;; ruleid: clojure-path-traversal
    (clojure.java.io/reader r)))

(defn safe []
  ;; ok: clojure-path-traversal
  (slurp "/tmp/fixed.txt")
  ;; ok: clojure-path-traversal
  (slurp "/tmp/other.txt")
  ;; ok: clojure-path-traversal
  (clojure.java.io/reader "/tmp/fixed.txt"))
